[riscv] Add support for RISC-V vector extension in arch

ita-sc opened the pull request:
First part of initial support for RISC-V vector extension.

Full pack of patches for support could be found in https://github.com/jfalcou/eve/pull/1716

With this patch `unit.arch` tests are fully passed.

jfalcou commented:
Hi, 
is using stuff like [Dockcross](https://github.com/dockcross/dockcross) viable for testing ?
It would simplify our job of settign up a proper test image.

DenisYaroshevskiy commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+  static constexpr std::size_t fundamental_cardinal = Size / 8 / sizeof(Type);
+
+  template<typename Type>
+  static constexpr std::size_t expected_cardinal = Size / 8 / sizeof(Type) * MaxLmul;
```
there is some confusion here.

I thought fundamental cardinal should be Size / 8 / sizeof(Type) / MaxLmul to represent negative multipliers

ita-sc replied:
To make this, cardinals should be type-dependent. Here I described how types are formed and why some negative multipliers can not be used with some element types https://github.com/jfalcou/eve/pull/1920#discussion_r1720966727. For this case let's assume you want to use `std::uint64_t` as element type. If your vlen is 64, you can not use part of vector register to represent even one 64 bit element, so no fractional LMUL for this element type available. There I use smallest vector register type that could fit for every element type.

ita-sc replied:
Oh, it seems that they are already type-dependent. Fixed.

DenisYaroshevskiy commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+    if( type_size == 4 ) return 2;
+    // for bigger types we can not use frac lmul.
+    // Return bigger than size of vector register value.
+    return Size + 1;
```
Why not optional?

ita-sc replied:
I thought it will be difficult to understand. Refactored.

DenisYaroshevskiy commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+    return std::bit_ceil(lmul);
+  }
+
+  template<arithmetic_scalar_value T> using m1_fixed = fixed<Size / 8 / sizeof(T)>;
```
this looks like the same code as on any other platform. Other platforms don't do have m1 fixed. what's this

ita-sc replied:
This is needed to construct wide that will use 1 vector register only. The is needed insert/extract values from RISC-V logical more effective. More detailed what I do with this here: https://github.com/jfalcou/eve/pull/1920#discussion_r1720966727

DenisYaroshevskiy commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  // return natural lmul if > 0, frac otherwise
+  template<typename Type> static consteval int getLMUL(size_t cardinal)
```
cardinal is always power of 2. seems like some code here is excessive.

ita-sc replied:
Not understand this part. We should anyway write some generic algorithm, I can not here filter cases when cardinal is not power of two. Should I add assert that cardinal is power of 2?

ita-sc replied:
Got it. Removed them

DenisYaroshevskiy reviewed (commented):
thank you for splitting. It's still quite big - we'll do in chunks anyways. Since information on rvv is not amazing, there is a lot we don't understand - there'll be some back and forth.

DenisYaroshevskiy commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,457 @@
+  static constexpr bool is_fp_v     = std::is_floating_point_v<Type>;
+  static constexpr bool is_signed_v = std::is_signed_v<Type>;
+
+  static consteval auto find_vint_mf8()
```
there was some description somewhere, to what all this `m` mean - can you share please? I don't understand the docs well

ita-sc replied:
RISC-V vector instrinsic types that are correspond for integer/float types, constructed with:
1)  the part that shows the underlying element type: `vint` - signed integer, `vuint` - unsigned integer, `vfloat` - float or double, depending on SEW
2) the part that shows `EEW` (effective element width) of one element - it can be 8, 16, 32, 64, with some restrictions, e.g. you can not have float with SEW=8. Note: I may sometimes refer to `SEW` - single element width, in terms of intrinsics they behave the same or almost the same, the difference appears on asm level, as there are some  vector instructions, that "hardcode" SEW to another value.  
3) the part that shows `EMUL`(effective vector register group multiplier) configuration (or what part of real vector register should be used to form type). Again, I can say `LMUL`(vector register group multiplier), and in terms of intrinsics this is the same, the difference again appear on asm level, as when you "hardcode" `SEW`, `LMUL` is changed too. There we have 
3.1) `mf8` - this is fractional usage 1/8 of real vector register
3.2) `mf4` - this is fractional usage 1/4 of real vector register
3.3) `mf2` - this is fractional usage 1/2 of real vector register
3.4) `m1` - usage of 1 real vector register
3.5) `m2` - usage of 2 real vector registers
3.6) `m4` - usage of 4 real vector registers
3.7) `m8` - usage of 8 real vector registers

Note, that some combinations are unsupported. For example, there is no type `vint32mf8_t`, as this will require having this type for every vlen, including vlen=64. But for this vlen we have 1/8 of vector register equals to 8, that can not fit int.

For supported types you can check https://github.com/riscv-non-isa/rvv-intrinsic-doc/releases/tag/v1.0-rc2 `Chapter 7. Type system`. Note: part `7.4. Tuple type` is not interested in context of this review.

ita-sc replied:
Done

jfalcou commented:
@ita-sc woudl you mind point em where to find info about the 1st part of the toolchain integration (the gcc-13 sysroot) ?

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+extract(wide<T, N> const& v, std::size_t i) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  auto OnFirstNeeded = __riscv_vslidedown_tu(v, v, i, N::value);
```
on_fist_needed

ita-sc replied:
Changed

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+  constexpr size_t sew              = sizeof(unsigned) * 8;
+  size_t           expected_element = i / sew;
+  auto             val              = extract(uint_wide, expected_element);
+  bool             to_ret           = std::bitset<sew>(val)[i % sew];
```
don't use std::bitset, it's pretty bad.

ita-sc replied:
Removed as now it uses store/load from stack.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+{
+  using u_m1_cardinal               = typename riscv_rvv_dyn_::m1_fixed<unsigned>;
+  using inter_type                  = wide<unsigned, u_m1_cardinal>;
+  auto             uint_wide        = bit_cast(v, as<inter_type> {});
```
I don't get this. Why are you bitcasting specifically to uint32_t

ita-sc replied:
This is comes due to restrictions of RISC-V vector ISA. For storing logical, I use rvv intrinsic type called mask. This is in fact bits, where n-th bit refer to the value of logical in n-th position. Hovewer, in RISC-V ISA I do not know any way to get n-th bit from mask, so I cast it to several unsigned, get interesting unsigned, and in it find the needed bit.

Let me show you example. Note: for simplification, I will use bellow casting to `unsigned char`, but in terms of effectiveness `unsigned` will be better. Suppose you have `logical<wide<int, 16>>=={1,0,1,0,0,1,0,1,0,1,0,1,1,01,0}`.  It will be stored in RISC-V vector register as:
`x....x1010010101011010`<sub>2</sub>, where `x` means that value is unknown. To get, for example, 10th bit, I do the following:
1) bitcast this vector to wide type (`unsigned char` is used for convince), that will occupy one vector register (note: there can not be used part of vector register, as for vector mask we occupy one full register). By the way, that is why I need `m1_fixed` - to get wide type that could be used to occupy only one vector register.
In result, for vlen=128, I will have `wide<unsigned char, 16>` with values:
<...`x`<sub>2</sub>,`x`<sub>2</sub>,`10100101`<sub>2</sub>,`01011010`<sub>2</sub>>
2) Get from casted `wide` scalar, where we have interesting for us bit (we still have no single instruction, but there we have some ways to do this), so we have `10100101`<sub>2</sub>
3) Get from resulted scalar interesting for us bit. There it will be I suppose `1`.

Also bellow I do the same thing when I need to insert bool into logical.

DenisYaroshevskiy replied:
This function seems very questionable. Just store the mask to the stack and return a bit from the array. We can think of something better later.

ita-sc replied:
Done. Though I suppose later I will create a separate review with this optimization.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+                extract(logical<wide<T, N>> const                &v, std::size_t i) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  using u_m1_cardinal               = typename riscv_rvv_dyn_::m1_fixed<unsigned>;
```
please use explicit types, unless you mean sizeness: std::uint32_t in this case.

ita-sc replied:
Seems fixed

DenisYaroshevskiy commented on test/unit/arch/top_bits.cpp:
```diff
@@ -38,12 +38,15 @@ TTS_CASE_TPL( "Check top bits raw type", eve::test::simd::all_types)
   using v_t         = eve::element_type_t<T>;
   using logical     = eve::logical<T>;
+  using c_t              = eve::cardinal_t<T>;
+  using rvv_logical_type = eve::logical<eve::wide<v_t, c_t>>;
```
this is just logical

ita-sc replied:
Fixed

jfalcou commented:
I am able to compile and run RISC-V random code using :

 + `riscv64-linux-gnu-g++-13 -static -march=rv64gcvzvl128b -std=c++20` using the Ubuntu RISC-V g++ 13 package
 + `qemu-riscv64 --cpu rv64,zba=true,zbb=true,v=true,vlen=128,vext_spec=v1.0`
 
 Could this be enough as a first CI target while clang settle down ?

ita-sc commented on cmake/toolchain/clang.rvv128.cmake:
```diff
@@ -0,0 +1,19 @@
+  message(FATAL_ERROR "expected to have RISCV_GCC in environment")
+endif()
+
+set(CMAKE_CXX_FLAGS "-O3 -march=rv64gcv -std=c++20 -mrvv-vector-bits=128 --sysroot=$ENV{RISCV_GCC}/sysroot --gcc-toolchain=$ENV{RISCV_GCC} --static --target=riscv64-unknown-linux-gnu ${EVE_OPTIONS}" )
```
For now I suppose better to start with vlen=128, later we could try to extend testing to bigger vlen. The main reason for this is testing time - even for vlen=128 compilation of unit testing for all cases quite long.

ita-sc commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+//================================================================================================
+template<std::size_t Size> struct rvv_abi_
+{
+  static constexpr std::size_t MaxLmul         = 8;
```
> From what I can see it's like the hardware will start using more registers

Yep, that is correct: when lmul > 1 in operation more physical vector registers are used.

> Is it beneficial in any way, over just unrolling n times?

For now I just optimized this in terms of number of executed instructions. I suppose by default it is expected that performing one vector operation with 8 registers is faster that performing 8 vector operations with 1 vector register, otherwise I see no reasons for adding this ability in ISA. Though this value seems to be hardware-dependent.

ita-sc commented:
> I am able to compile and run RISC-V random code using :
> 
> * `riscv64-linux-gnu-g++-13 -static -march=rv64gcvzvl128b -std=c++20` using the Ubuntu RISC-V g++ 13 package
> * `qemu-riscv64 --cpu rv64,zba=true,zbb=true,v=true,vlen=128,vext_spec=v1.0`
> 
> Could this be enough as a first CI target while clang settle down ?

You mean simple code, without attributes for sized types? Unfortunately, in can not be used, as without them, vector types can not be put in structures (as they are sizeless by default).

I see that in gcc-14.2 there is some kind of support for sized vector types, but I need to check if requirement for this mr are supported.

As first stage, we can use even clang-19, the main reason for update is that without patch you can not create `logical<wide<T, N>>`, where N < 8.

And for full support in clang, we need quite new llvm: the patch with support required functionality in clang was merged last week (https://github.com/llvm/llvm-project/commit/635d20e9e72e25966b196bb012a90e750ae00c6d)

> @ita-sc woudl you mind point em where to find info about the 1st part of the toolchain integration (the gcc-13 sysroot) ?

I usually build it myself, or use prebuild one (It seems you can use, for example https://syntacore.com/tools/development-tools - there path to folder `riscv-gcc` should be used as `RISCV_GCC`). If you anyway install of riscv-gcc-13 in your system, I suppose I can change scripts so you need to export variables to installed packages.

ita-sc commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+struct riscv_rvv_dyn_ : rvv_abi_<__riscv_v_fixed_vlen>
+{};
+#else
+struct riscv_rvv_dyn_ : rvv_abi_<1>
```
Currently, if vlen is not specified (that is no `-mrvv-vector-bits` is set), compilation results in an error. This happens because constexprs are unable to be evaluated properly. It seems that emulation does not work for RISC-V even without my changes. 

My idea was following: I wanted to try an set vlen to some non-sensible value in a hope that this will force template magic to enforce emulation. Now, I see that this attempt failed because of the following errors (again, this happens without my patches):

```
error: invalid operands to binary expression ('const ::eve::cpu_' and const asimd_')
...
```

For now added instead of this `#error` with description, for these compilation failures to be a little less cryptic. 

By the way, do you maybe have any recommendations on how to make emulation work?

ita-sc replied:
> I need more context. Does it fail when you try to compile without any changes? Where?

I see currently errors when I try to compile for RISC-V on current main branch (`e16c5bf2e43c8c09e2947c466d9c85eafbce3ec5`). To reproduce:
1) Create a toolchain file `cmake/toolchain/gcc.riscv.cmake
` (assume you have any riscv-gcc installed):
```
set(CMAKE_SYSTEM_NAME Linux     )
set(CMAKE_SYSTEM_PROCESSOR riscv64  )

set(CMAKE_C_COMPILER     riscv64-linux-gnu-gcc  )
set(CMAKE_CXX_COMPILER   riscv64-linux-gnu-g++  )
set(CMAKE_CXX_FLAGS     "-O3 -march=rv64gcv -std=c++20 -W ${EVE_OPTIONS}" )
```
2) Run config:
```
cmake -B build \
  -S . -G Ninja\
  -DCMAKE_TOOLCHAIN_FILE=cmake/toolchain/gcc.riscv.cmake
```

3) Run build (for example, of `unit.arch`):
```
cmake --build build --parallel 4 --target unit.arch.exe
```

On this set up, I see a lot of errors, the first one is:
```
.../include/eve/forward.hpp:42:80: error: no match for 'operator>=' (operand types are 'const eve::cpu_' and 'const eve::avx512_')
   42 | auto EVE_FORCEINLINE to_logical(C c, eve::as<T>) noexcept requires(current_api >= avx512);
      |                                                                    ~~~~~~~~~~~ ^~ ~~~~~~
      |                                                                    |              |
      |                                                                    |              const eve::avx512_
      |                                                                    const eve::cpu_
```


> See how predefs are done for arm-sve:

Yep, I suppose now they are working the same for RISC-V.

ita-sc commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,115 @@
+//================================================================================================
+// ABI tags for all RISCV bits SIMD registers
+//================================================================================================
+template<std::size_t Size> struct rvv_abi_
```
Rename to `VectorRegisterBitSize`

jfalcou commented:
> > I am able to compile and run RISC-V random code using :
> > 
> > * `riscv64-linux-gnu-g++-13 -static -march=rv64gcvzvl128b -std=c++20` using the Ubuntu RISC-V g++ 13 package
> > * `qemu-riscv64 --cpu rv64,zba=true,zbb=true,v=true,vlen=128,vext_spec=v1.0`
> > 
> > Could this be enough as a first CI target while clang settle down ?
> 
> You mean simple code, without attributes for sized types? Unfortunately, in can not be used, as without them, vector types can not be put in structures (as they are sizeless by default).

Of course :E 

> And for full support in clang, we need quite new llvm: the patch with support required functionality in clang was merged last week ([llvm/llvm-project@635d20e](https://github.com/llvm/llvm-project/commit/635d20e9e72e25966b196bb012a90e750ae00c6d))

I am OK with getting a trunk clang in our CI image no problem.

> I usually build it myself, or use prebuild one (It seems you can use, for example https://syntacore.com/tools/development-tools - there path to folder `riscv-gcc` should be used as `RISCV_GCC`). If you anyway have install of riscv-gcc-13 in your system, I suppose I can change scripts so you need to export variables to installed packages.

Let's do that thanks

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+  using Wide  = logical<wide<T, N>>;
+  using abi_t = typename Wide::abi_type;
+
+  if constexpr( has_bundle_abi_v<Wide> )
```
logical should never have a bundle abi, this seems like dead code to me.

DenisYaroshevskiy commented:
FYI - if you want to - we can set some time to pair up and review this on a call, so that we don't have to communicate asynchronously.



You can reach me on denis.yaroshevskij@gmail.com, dyaroshev on twitter or cpplang slack.

ita-sc commented:
>> I usually build it myself, or use prebuild one (It seems you can use, for example https://syntacore.com/tools/development-tools - there path to folder riscv-gcc should be used as RISCV_GCC). If you anyway have install of riscv-gcc-13 in your system, I suppose I can change scripts so you need to export variables to installed packages.

> Let's do that thanks

Refactored in the same way you use clang for arm - you need to add to your `LD_LIBARY_PATH` path to RISC-V libs (seems `/usr/riscv64-linux-gnu/lib/`)

ita-sc commented:
> You can reach me on denis.yaroshevskij@gmail.com, dyaroshev on twitter or cpplang slack.

I think this is a great idea, I've sent you an email.

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/if_else.hpp:
```diff
@@ -0,0 +1,63 @@
+{
+  auto fwide = make(as<wide<T, N>> {}, vf);
+  return if_else(c, vt, fwide);
+}
```
This should work out of the box, I'm surprised you need it.

ita-sc replied:
Otherwise I have recursion: for implementing function `to_bits` and similar I use `if_else` with constants.

DenisYaroshevskiy replied:
The decisions that lead to this seem incorrect.

DenisYaroshevskiy replied:
Make an if constexpr chain plz

ita-sc replied:
I'm not sure I understand you correctly. To be in context, here is how used target-dependent hooks on other archs in EVE:
arm neon:
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

arm sve:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L17

* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L29

ppc:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L18
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L28

x86:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L26
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

Do you imply that for RISC-V I should implement a different approach? Like to have the only one function:
```
template<logical_simd_value Logical, Value VT, Value VF>
requires(std::same_as<element_type_t<VT>, element_type_t<VF>> && std::same_as<element_type<typename Logical::mask_type>, element_type_t<VT>> && rvv_abi<...> && /*same cardinal, if Values are wide */)
auto if_else(..., Logical logical,, VT vt, VF vf) {
if constepxr (std::same_as<Logical, VT> && std::same_as<Logical, VF>) {...}
else if constexpr (...)
....
}
```

Also note that combining all if_else functions will require at least
quite a huge `requires` block, that may not work, as it is not working
for friend functions: compiler will not be able to distinguish
common `if_else` from riscv-specific, as both will match.

ita-sc replied:
I'm not sure I understand you correctly. To be in context, here is how used target-dependent hooks on other archs in EVE:
arm neon:
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

arm sve:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L17

* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L29

ppc:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L18
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L28

x86:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L26
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

Do you imply that for RISC-V I should implement a different approach? Like to have the only one function:
```
template<logical_simd_value Logical, Value VT, Value VF>
requires(std::same_as<element_type_t<VT>, element_type_t<VF>> && std::same_as<element_type<typename Logical::mask_type>, element_type_t<VT>> && rvv_abi<...> && /*same cardinal, if Values are wide */)
auto if_else(..., Logical logical,, VT vt, VF vf) {
if constepxr (std::same_as<Logical, VT> && std::same_as<Logical, VF>) {...}
else if constexpr (...)
....
}
```

Also note that combining all if_else functions will require at least
quite a huge `requires` block, that may not work, as it is not working
for friend functions: compiler will not be able to distinguish
common `if_else` from riscv-specific, as both will match.

DenisYaroshevskiy replied:
I see - scalar overloads.

There is a problem though. 

You have 3 of scalar overloads and you are not always calling the correct one. 
The problem comes in when the scalar type is not exactly matching T.
Then you would do a broadcast, which your code tries to avoid.

I don't see you escaping having to

```
template <logical_simd_value L, arithmetic_value T, arithmetic_value U>
auto if_else_(L, T, U);
```

If you want to overload: `wide, wide`, `wide, scalar`, `scalar, wide` - fine, but you still need to figure out casts. 

FYI: this is how you call generic version: https://github.com/jfalcou/eve/blob/89d8ab2092984c3511ecdefeed0ff9c4d96c4fdd/include/eve/module/core/compress/simd/x86/compress.hpp#L20

ita-sc replied:
> I don't see you escaping having to
template <logical_simd_value L, arithmetic_value T, arithmetic_value U>
auto if_else_(L, T, U);

Got you. Done, now for scalars we call `static_cast` to the right type, and call `if_else` again

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/if_else.hpp:
```diff
@@ -0,0 +1,63 @@
+#include <eve/detail/function/simd/riscv/make.hpp>
+#include <eve/detail/implementation.hpp>
+
+namespace eve::detail
```
I see that in other files here wide and logical are passed by reference, but that's a mistake - they are always passed by value. 

Also - you don't need to put noexcepts. It doesn't matter one way or the other but i generally don't.

ita-sc replied:
> I see that in other files here wide and logical are passed by reference, but that's a mistake - they are always passed by value.

Fixed. By the way, it seems in arm SVE you use const refs:

https://github.com/jfalcou/eve/blob/main/include/eve/detail/function/simd/arm/sve/combine.hpp#L19

> you don't need to put noexcepts

Well, AFAIK for `noexcept` functions compiler could be more aggressive with optimizations. Here it should not be the real point though, but I prefer for now to keep them, if you are not against.

jfalcou replied:
noexcept only make code better when used in place where noexcept quality of function is important : move semantic, copy etc. Mostly in standard container. We do mark them noexcept because they are and mayeb , one day, maybe, it'll have an impact. But the code itself of the function don't really benefit from this.

DenisYaroshevskiy replied:
noexcept is completly irrelevant for force inline funtions. w/e.

ita-sc commented on include/eve/arch/abi_of.hpp:
```diff
@@ -65,6 +65,11 @@ namespace eve
+        else if constexpr( spy::simd_instruction_set == spy::rvv_
+                           && spy::simd_instruction_set.width >= 64 )
+        {
+          return riscv_rvv_dyn_ {};
```
What I mean is: instead of different kind of tags for different vlen (as it is done in ARM SVE), I use the only one tag: `riscv_rvv_dyn_`. Dynamic in this place means that the same code should work if you compile with vlen equals 64, 128, 256... (Though for now I'm testing only variant with vlen=128).

How about `riscv_rvv_vlen_`?

jfalcou commented on include/eve/arch/abi_of.hpp:
```diff
@@ -65,6 +65,11 @@ namespace eve
           }
         }
+        else if constexpr( spy::simd_instruction_set == spy::rvv_
+                           && spy::simd_instruction_set.width >= 64 )
```
we need to amend SPY for RISC-V first. I can look at that

jfalcou commented on include/eve/arch/as_register.hpp:
```diff
@@ -10,9 +10,10 @@
-#include <eve/arch/ppc/as_register.hpp>
-#include <eve/arch/arm/sve/as_register.hpp>
-#include <eve/arch/arm/neon/as_register.hpp>
+#  include <eve/arch/arm/neon/as_register.hpp>
```
this is noise.

ita-sc replied:
Accidental change by clang-format. Fixed.

jfalcou commented on include/eve/arch/riscv/spec.hpp:
```diff
@@ -0,0 +1,39 @@
+//==================================================================================================
+// Register count
+//==================================================================================================
+#if defined(EVE_HW_RISCV_SVE)
```
I don't think this is the correct macro

ita-sc replied:
Do you mean that I should use `EVE_INCLUDE_RISCV_HEADER`? Fixed, and removed for now define `EVE_HW_RISCV_SVE` define, as it is not used now.

jfalcou commented on include/eve/arch/spec.hpp:
```diff
@@ -8,10 +8,11 @@
-#  include <eve/arch/ppc/spec.hpp>
-#  include <eve/arch/arm/sve/spec.hpp>
 #  include <eve/arch/arm/neon/spec.hpp>
+#  include <eve/arch/arm/sve/spec.hpp>
```
Again, there is no need to reorder those

ita-sc replied:
Accidental change by clang-format. Fixed.

jfalcou commented on include/eve/arch/tags.hpp:
```diff
@@ -7,8 +7,9 @@
 
+#include <eve/arch/arm/neon/tags.hpp>
+#include <eve/arch/arm/sve/tags.hpp>
 #include <eve/arch/cpu/tags.hpp>
```
cpu tags must be first. Don't reorder lines for no reason.

ita-sc replied:
Accidental fix by clang-format. Fixed.

jfalcou commented on include/eve/detail/spy.hpp:
```diff
@@ -10,21 +10,24 @@
-                    , wasm_ = 40
-                    };
-  template<archs Arch> struct arch_info
+enum class archs {
```
SPY is an external dependencies, reformatting it with EVE clang format is not necessary.
Also, care to put an issue at SPY for RISC-V support ?

jfalcou replied:
Good, I'll take care of that

jfalcou commented on include/eve/detail/spy.hpp:
```diff
@@ -872,48 +886,80 @@ namespace avx512
+  vmx_2_03_  = 2203,
+  vmx_2_05_  = 2205,
+  vmx_2_06_  = 2206,
+  vmx_2_07_  = 2207,
```
again, unrequried reformatting

ita-sc replied:
Fixed

DenisYaroshevskiy commented on test/test.hpp:
```diff
@@ -278,7 +281,8 @@ namespace tts
     // For emulated type, there is no such gap so we don't do anything
-    if constexpr( (W::size() < eve::fundamental_cardinal_v<v_t>) && !eve::has_emulated_abi_v<W> )
+    if constexpr( (W::size() < eve::fundamental_cardinal_v<v_t>)&&!eve::has_emulated_abi_v<W>
+                  && eve::current_api != eve::rvv )
```
I'm not sure what's happening here and it doesn't strike me as correct.

ita-sc replied:
Yep, you are right - after change of `is_full` it RISC-V tags this change is not needed, so removing this.

Earlier I supposed that `is_full` should refer to filling max available physical registers. Because of this, types corresponding for `W data` and ` ftype these` will not be the same, so we can not construct `ftype` from `storage` of `data`.

Now, as `is_full` implemented in terms of intrinsics types,  we are able to construct `ftype` with `storage`.

DenisYaroshevskiy commented on test/test.hpp:
```diff
@@ -294,6 +298,24 @@ namespace tts
+    else if constexpr( (W::size() < eve::fundamental_cardinal_v<v_t>)&&!eve::has_emulated_abi_v<W>
+                       && eve::current_api == eve::rvv )
+    {
+      // for riscv we can not just cast ftypes from storage type, so use bitcast.
```
that seems like btcast no working.

ita-sc replied:
Removed

DenisYaroshevskiy commented on include/eve/detail/spy.hpp:
```diff
@@ -13,7 +13,7 @@ namespace spy::detail
                     , x86_ = 10, amd64_ = 11
                     , ppc_ = 20, arm_ = 30
-                    , wasm_ = 40
+                    , wasm_ = 40, riscv_ = 50
```
I can't even remotely comment on spy changes, @jfalcou - this is on you.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/to_logical.hpp:
```diff
@@ -0,0 +1,26 @@
+requires rvv_abi<abi_t<T, N>>
+{
+  return v != static_cast<T>(0);
+}
```
I'm surprised that you need this, but it's a to_logical problem, not your problem.

jfalcou commented:
OK so I have a working clang trunk for RISC-V. I'll be testing it this week-end on SPY and Kumi to see if it's OK.
Once done, I'll pass you a gist that you'll add to this PR to make your core run on our CI.

Thanks for all your pointers !

camel-cdr commented:
I'd try to maximise LMUL without causing register spills.
If I understand eve correctly then it's a generic SIMD abstraction, and since you don't know the register preasure of code using eve, defaulting to LMUL=1 is the safest option. Considering that eve supports sse and avx2 which both only have 16 registers, going with LMUL=2, which gives you 16 registers would also be possible.

DenisYaroshevskiy commented:
@ita-sc  - chosing better default won't be really possible at this stage, eve is designed around maximum being the right default. We'll fix that later.

ita-sc commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,126 @@
+struct riscv_rvv_dyn_ : rvv_abi_<__riscv_v_fixed_vlen>
+{};
+#else
+struct riscv_rvv_dyn_ : rvv_abi_<1>
```
Done

ita-sc commented on include/eve/arch/riscv/top_bits.hpp:
```diff
@@ -0,0 +1,146 @@
+  EVE_FORCEINLINE constexpr bool operator==(top_bits const& x) const
+  {
+    auto neq_res = __riscv_vmxor(storage, x.storage, static_size);
+    return __riscv_vcpop(neq_res, static_size) == 0;
```
`operator==` for `logical` returns `logical`. I suppose `operator==` should return just `bool`, so `operator==` is not enough.

I'm using `operator!=` and than I expect, that `top_bits` from that value should be false (that is no `true` values in storage `logical`). 

Refactored. Could you take a look now?

ita-sc commented on include/eve/detail/function/simd/riscv/bitmask.hpp:
```diff
@@ -0,0 +1,46 @@
+requires rvv_abi<abi_t<T, N>>
+{
+  // As return by arm.
+  using int_type       = as_integer_t<T, std::uint32_t>;
```
Seems working. Fixed.

ita-sc commented on include/eve/detail/function/simd/riscv/combine.hpp:
```diff
@@ -0,0 +1,59 @@
+    that.storage().assign_parts(l, h);
+    return that;
+  }
+  else { return to_logical(eve::combine(l.mask(), h.mask())); }
```
Added

ita-sc commented on include/eve/detail/function/simd/riscv/compounds.hpp:
```diff
@@ -0,0 +1,321 @@
+  auto           y = static_cast<T>(other);
+  if constexpr( match(c, category::integer_) ) return self = __riscv_vmul(self, y, N::value);
+  else if constexpr( match(c, category::float_) ) return self = __riscv_vfmul(self, y, N::value);
+}
```
Yep, combined these functions

ita-sc commented on include/eve/detail/function/simd/riscv/compounds.hpp:
```diff
@@ -0,0 +1,321 @@
+  }
+  else
+  {
+    using scalar_t           = as_integer_t<T, std::uint32_t>;
```
For now I suppose I can even delete these functions, as for now target is to make `unit.arch` tests passed. I will add them in next review with changes. I think I have an idea how to merge them.

ita-sc commented on include/eve/detail/function/simd/riscv/friends.hpp:
```diff
@@ -0,0 +1,282 @@
+  auto rhs_cast = static_cast<T>(rhs);
+  if constexpr( std::is_signed_v<T> ) return __riscv_vmslt(lhs, rhs_cast, N::value);
+  else return __riscv_vmsltu(lhs, rhs_cast, N::value);
+}
```
Merging them in one `self_*` function is not possible. For example let's check `self_less`. In common overload it has:
```
EVE_FORCEINLINE auto self_less(Wide const& v,Wide const& w) noexcept
```

So any riscv-specific overload that uses
```
self_less(wide<T, N> lhs, U rhs)
```
seems for any combination of requirements will be ambiguous with the common one.
As it seems not possible, I've created a `self_*_impl` functions with this logic to remove code duplication, and saved two interfaces.

ita-sc commented on include/eve/detail/function/simd/riscv/friends.hpp:
```diff
@@ -0,0 +1,282 @@
+    auto                casted_w = bit_cast(w, as<logical<wide<T, N>>> {});
+    logical<wide<T, N>> to_ret   = __riscv_vmand(v, casted_w, N::value);
+    return to_ret;
+  }
```
Here I've added this, as in logical's that I've created for RISC-V are different. Suppose you want to fully fill `char` `wide` with vlen=128 and lmul = 8. It will be `wide<char, 128>`. But for `int`, the maximum `wide`, that could be placed on physical registers is `wide<int, 32>` (128 / 32 * 8). If we have logical, that could be effectively work with this type of wide, even if it can actually "fit" into native vector register mask type, I split them. Otherwise, operations for aggregated `wide` with predicate will not be cheap (I will have to construct real mask types on place).

ita-sc commented on include/eve/detail/function/simd/riscv/load.hpp:
```diff
@@ -0,0 +1,92 @@
+    return __riscv_vle32_tumu(mask, zero_init, p, N::value);
+  else if constexpr( match(c, category::size64_) )
+    return __riscv_vle64_tumu(mask, zero_init, p, N::value);
+}
```
Done

ita-sc commented on include/eve/detail/function/simd/riscv/load.hpp:
```diff
@@ -0,0 +1,92 @@
+  if constexpr( C::has_alternative )
+  {
+    auto res = load(drop_alternative(cond), s, tgt, p);
+    return eve::replace_ignored(res, cond, cond.alternative);
```
Yep, done

ita-sc commented on include/eve/detail/function/simd/riscv/load.hpp:
```diff
@@ -0,0 +1,92 @@
+{
+  auto const c1    = map_alternative(cond, [](auto alt) { return alt.mask(); });
+  auto const block = load(c1, safe, eve::as<wide<T, N>> {}, ptr_cast<T const>(ptr));
+  return block != 0;
```
Done

ita-sc commented on include/eve/arch/riscv/top_bits.hpp:
```diff
@@ -0,0 +1,146 @@
+  EVE_FORCEINLINE constexpr auto as_int() const
+  requires(static_bits_size <= 64)
+  {
+    constexpr size_t               size   = __riscv_v_fixed_vlen / 8;
```
Done. I had to add additional logic for cases when `static_bit_size < 8`

ita-sc commented:
Seems I've addressed all comments. Also in separate commit I change places, where for RISC-V I need to use `simd_cast` instead of `bit_cast`. RISC-V functions for `bit_cast` now can not change the size of underlying type, this work goes to `simd_cast`. @DenisYaroshevskiy, @jfalcou any more comments?

DenisYaroshevskiy commented:
Sorry for being slow. Didn't have time to re-review again. Let me do a bit.

We are blocked on the ci supporting this update.
You will also have a small and annoying rebase coming, when @SadiinsoSnowfall finishes migrating operators (sorry about that) but it'll make the code simpler.

DenisYaroshevskiy commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,497 @@
+
+namespace eve
+{
+template<typename Type, typename Size, rvv_abi ABI> struct as_register<Type, Size, ABI>
```
I think, now that I understand it, I can simplify this code a bit. 
However, I would like to compile the code I'd suggest first.

Is there a way I can write rvv simd code on godbolt?

ita-sc replied:
I suppose you can. Here is an example: https://godbolt.org/z/ThfEqhc59  (just vector add with lmul=1 and scalar type `std::uint64_t`). The main idea here is to add to riscv clang `-march=rv64gcv -mrvv-vector-bits=128`.

By the way, here I use load with mask, as in EVE, but actually I can perform nomask load. This should be faster, but will result in much more code for review (as I will need to write separate intrinsic for each combination of type/lmul). I suggest to make this optimization in separate review.

DenisYaroshevskiy replied:
Here is what I have: https://godbolt.org/z/GcEz5v95j
Feel vindicated, I had no idea that they decided to support none of the normal c++ things, we have to some macro.

I think having one long table is nice for me.

ita-sc replied:
I've used some defines from your code. Unfortunately, your calculation of LMUL not type-dependent, so it can not be used ( as some types, like `int`, can not be used with frac lmul 1/4 or 1/8 for example), so I still use the one that is placed in tags (I need the these calculations for `logical` too, so placing it in `as_register` is not convenient). 

> one long table is nice for me

I still prefer to use tree-like function calls, so, if you do not mind, I prefer to keep for now structure of getting registers, as set of short functions looks better (and will be easier to debug) in my opinion. If you think that one long function will be better, I will replace this.

By the way, in the original review (before deleting all the macros), I used more interesting macros, with concatenation, so you will not need to use for example `4` (as LMUL value) several times (and `4`). May be return that style, as we already return defines?

DenisYaroshevskiy replied:
> Uortunately, your calculation of LMUL not type-dependent, so it can not be use so it can not be used
Yes, it is: 

```
  constexpr std::ptrdiff_t m1_len = __riscv_v_fixed_vlen;
  constexpr std::ptrdiff_t min_len = m1_len * sizeof(T) / 8;
  constexpr std::ptrdiff_t reg_len = std::max(min_len, N);
```



> set of short functions looks better
I don't agree with you here but I can read it now, so w/e.

> more interesting macros
No, that one is a no

ita-sc replied:
Doesn't saw that you use `constexpr std::ptrdiff_t min_len = m1_len * sizeof(T) / 8;`. It seems by this you can get expected frac lmul. I've refactored code to the same manner. Could you take a look? I still need to change code, as `std::max(min_len, N);` looks like error (max between cardinal and vector register size).

DenisYaroshevskiy commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,497 @@
+    static_assert(width <= __riscv_v_fixed_vlen * ABI::MaxLmul,
+                  "[eve riscv] - Type is not usable in SIMD register (too big)");
+    if constexpr( is_fp_v ) return find_vfloat();
+    else if constexpr( !is_fp_v && is_signed_v ) return find_vint();
```
you don't need to repeat conditions, if constexpr respects else fine

DenisYaroshevskiy commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,497 @@
+    constexpr auto width = sizeof(Type) * Size::value * 8;
+
+    static_assert(width <= __riscv_v_fixed_vlen * ABI::MaxLmul,
+                  "[eve riscv] - Type is not usable in SIMD register (too big)");
```
this is fairly internal - I wouldn't worry about it. We have a list of supported types already.

ita-sc replied:
I prefer to keep this to be sure that `as_logical_register` is in sync with `as_register`. If you wish, I will delete this.

ita-sc replied:
Done

DenisYaroshevskiy commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,303 @@
+  static_assert(!std::is_void_v<type>,
+                "[eve riscv] - Type is not usable as logical(mask) SIMD register");
+};
+}
```
Undef macros

ita-sc replied:
I have undef upper:

https://github.com/jfalcou/eve/pull/1920/files#diff-83b7e939c55478525ce4181d3250fef27c4c5482d56eee09518669a4c5d215adR248

DenisYaroshevskiy commented on include/eve/arch/riscv/rvv_common_masks.hpp:
```diff
@@ -0,0 +1,54 @@
+EVE_FORCEINLINE logical<wide<T, N>>
+                rvv_true()
+{
+  static constexpr auto   lmul  = riscv_::getLMUL<T>(N::value);
```
there should just be 

```
rvv_lmul_v<T>
```

@jfalcou  - where would you put it.

ita-sc replied:
I keep it in tags, as I though tags are used to have some kind of wrapping for real vector registers. Like, it is already have `fundamental_cardinal`, `expected_cardinal`, `is_full` and so on.

DenisYaroshevskiy commented on include/eve/arch/riscv/rvv_common_masks.hpp:
```diff
@@ -0,0 +1,54 @@
+{
+  static constexpr auto   lmul  = riscv_::getLMUL<T>(N::value);
+  static constexpr size_t size  = sizeof(T) * 8;
+  static constexpr size_t ratio = lmul > 0 ? size / lmul : size * (-lmul);
```
Same, we shouldn't redo them over and over. There shuld be a 

```
template <simd_value T>
constexpr std::ptrdiff_t rvv_logical_ratio_v = ...
```

or smth

I'm also not sure what raio is. why ratio?

ita-sc replied:
Earlier it was something like `bit_size` (I earlier review I was asked to rename this). In terms of mask intrinsic types, this is just the number in type. E.g. `vbool64_t` - we get 64. I get it as `SEW/LMUL`. This could represent the number of bits that this mask is stores (you get the number of bits by `vlen/number`).

Also in rvv intrinsics documentation they use ratio too: `Mask types have the ratio that is derived ...`

For now named function `getLogicalRatio` and placed it in tags.

DenisYaroshevskiy commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,96 @@
+    auto           lmul         = getLMUL<Type>(cardinal);
+    constexpr auto element_size = sizeof(Type) * 8;
+    return lmul > 0 ? element_size / lmul : element_size * (-lmul);
+  }
```
eve uses `_` notations.

This is what I suggest.

Similar to https://github.com/jfalcou/eve/blob/main/include/eve/arch/arm/sve/sve_true.hpp

Create `eve/arch/riscv/ratios.hpp` or smth

in that file:

```
template <simd_value T>
constexpr rvv_lmul_v = []{
  // ...
}();

template <logical_simd_value T>
constexpr rvv_logical_ratio_v = []{
  // ...
}();
```

most places - use those

DenisYaroshevskiy commented on include/eve/arch/riscv/top_bits.hpp:
```diff
@@ -0,0 +1,145 @@
+
+  static constexpr auto half_size = (static_size / 2 > 0) ? static_size / 2 : 1;
+  using half_logical              = logical<wide<scalar_type, eve::fixed<half_size>>>;
+  using storage_type              = logical<wide<scalar_type, eve::fixed<static_size>>>;
```
just logical_type

ita-sc replied:
Done

DenisYaroshevskiy commented on include/eve/arch/riscv/top_bits.hpp:
```diff
@@ -0,0 +1,145 @@
+    {
+      // if static_bit_size < 8, we have some dirty bits in resulted unsigned,
+      // so cleanning them.
+      u_type mask = (1 << static_bits_size ) - 1;
```
this doesm't work for all values of static bits size. You can shift by 64.
Use  detail::set_lower_n_bits<u_type>

ita-sc replied:
Done

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/bit_cast.hpp:
```diff
@@ -0,0 +1,314 @@
+//==================================================================================================
```
seems like this should've migrated to simd_cast.

ita-sc replied:
Not all: if lmul values for wide matches, we can use simple bitcasts, no changes for size of underlying types.

DenisYaroshevskiy replied:
then memcpy should work, you don't need the casts. The compiler should do the right thing.
We don't do this for other platfroms. Maybe we should but we don't. So let's not here.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/bitmask.hpp:
```diff
@@ -0,0 +1,39 @@
+requires rvv_abi<abi_t<T, N>>
+{
+  using uint_type = as_uinteger_t<T>;
+  return eve::if_else(p, static_cast<uint_type>(0), static_cast<uint_type>(0));
```
this looks like a bug. Is this code used?

ita-sc replied:
Fixed. Seems was broken during one of updates.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/combine.hpp:
```diff
@@ -0,0 +1,63 @@
+    auto           wider_l        = bit_cast(l, as<that_t> {});
+    auto           wider_h        = bit_cast(h, as<that_t> {});
+    constexpr auto shift_size     = N::value;
+    that_t         wider_h_placed = __riscv_vslideup(wider_h, wider_h, shift_size, combined_vl);
```
can't you just slide the l,h in position, why ifs? or doesn't work

ita-sc replied:
In asm you can just move physical registers where you need (if lmul >=1), but unfortunately, in intrinsics interface there is no way to put l,h. And for cases when lmul <1 you will still use this code.

DenisYaroshevskiy replied:
fine for now but if I come up with a way to do this better, before we merge this, I will request a change.

ita-sc commented on include/eve/arch/riscv/rvv_common_masks.hpp:
```diff
@@ -0,0 +1,50 @@
+  Vals[i]                      = 1;
+  wide<T, N> Vector(Vals.begin(), Vals.end());
+  return Vector != static_cast<T>(0);
+}
```
Not a lot: just in one place to insert value to `wide`. In future I guess we can even optimize this method(just like in insert/extract for logical - at first cast to `wide<int, 4>`, and change one int).

> return {[i](int j, int) { return j == i; }};

Done

ita-sc commented on include/eve/arch/riscv/rvv_utils.hpp:
```diff
@@ -0,0 +1,37 @@
+    std::ptrdiff_t           expected_len = sizeof(scalar_type) * 8 * cardinal::value;
+    std::ptrdiff_t           reg_len      = std::max(min_len, expected_len);
+    if( reg_len == m1_len ) return 1;
+    else if( reg_len > m1_len ) return static_cast<int>(reg_len / m1_len);
```
Ok, done

ita-sc commented on test/test.hpp:
```diff
@@ -158,6 +158,9 @@ namespace eve::test
         case 512: return {0,8,7,0,6,0,0,0,5};
+        case 1024: return {0, 6, 5, 0, 4, 0, 0, 0, 3};
+        // Unfortunately, this is too long on simulation
+        // case 1024:  return {0,9,8,0,7,0,0,0,6};
```
For `unit.arch` we do not have problems in timing. I have them when I run tests from `unit.algorithms` and `unit.api` lists. May be you have some way to filter cardinals for this folders?

As for now we test `unit.arch`, I suppose I can remove this comment, at least for now, should I?

DenisYaroshevskiy replied:
I don't really know this file, @jfalcou will review this.

I don't think we have good options, unfortunately.

I see two:

1. smth like what you did - where you just for now disable all large registers.
2. in the tests that time out - put a:

```
if constexpr (eve::current_api >= eve::rvv && T::size() > 32) {
   TTS_PASS();
   return;
}
```

And then we go clean those up at some point.
The second one seems slightly better, but it really depends on how many tests are exploding in time.

I know one thing - we should do up to 256 byte registers tested.

Everything else - up to @jfalcou

ita-sc replied:
For now set the expected cardinals (` case 1024:  return {0,9,8,0,7,0,0,0,6};`). I suppose we can return to this question as I will create MR with support for `unit.algorithms` or `unit.api`.

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/simd_cast.hpp:
```diff
@@ -0,0 +1,183 @@
+requires rvv_abi<abi_t<T, N>>
+{
+  constexpr auto c = categorize<wide<T, typename N::combined_type>>();
+  static_assert(N::combined_type::value != N::value);
```
Combined type should always be twice as big, I'm not sure what's that about.
Am I missing smth?

ita-sc replied:
> Am I missing smth?

Nop. I've added assert for to be sure. Removed

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/simd_cast.hpp:
```diff
@@ -0,0 +1,183 @@
+  else if constexpr( match(c, category::int32) )
+  {
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_i32m1(a);
+    else static_assert(false, "[eve riscv] Unsupported");
```
Let's replace the word `Unsupported` with `sanity check` - so that the user sees that we have an internal issue, if this happens.

ita-sc replied:
Yep, done

ita-sc replied:
Even removed for now, as `unit.arch` do not test this functionality. Will add with `sanity check` in future reviews

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/simd_cast.hpp:
```diff
@@ -0,0 +1,183 @@
+  else static_assert(false, "[eve riscv] Unsupported");
+}
+
+// Logical-wide conversions.
```
This seems like that rvv special magic to bit convert logicals to wides.

a) please leave a comment this is a special logic for rvv and that's what you are doing here.
b) I'm extremly confused why this would get called. You don't change the `enabled_for` interface to `simd_cast`. How is this called? Is it tested? https://github.com/jfalcou/eve/blob/89d8ab2092984c3511ecdefeed0ff9c4d96c4fdd/include/eve/module/core/regular/simd_cast.hpp#L70

ita-sc replied:
a) added
b) As for now testing is only `unit.arch`, we test only casting between logicals do the work good. 

Remove for now `simd_cast` support for different lmul sizes, as tests for them seems to be in different testpack.

> enabled_for

I do not need to change anything there, as there for logical already have cast between logical available:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/simd_cast.hpp#L75

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/simd_cast.hpp:
```diff
@@ -0,0 +1,183 @@
+
+template<plain_scalar_value T, typename N>
+EVE_FORCEINLINE wide<T, typename N::combined_type>
+                riscv_lmul_ext(wide<T, N> a) noexcept
```
Are you running clang format? I'm pretty sure it'd get reformatted.

It's ok - if you don't, then there will just be a reformatting patch later.

ita-sc replied:
Yep, I will run git-clang-format later

ita-sc replied:
Formatted

DenisYaroshevskiy commented on include/eve/module/core/regular/impl/simd/riscv/simd_cast.hpp:
```diff
@@ -0,0 +1,183 @@
+  }
+  else
+  {
+    using u_m1_cardinal  = typename riscv_::m1_fixed<std::uint32_t>;
```
You can probably have `using rvv_m1_wide = ` somewhere.

ita-sc replied:
Done. Moved this to rvv_utils

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/slice.hpp:
```diff
@@ -0,0 +1,187 @@
+
+template<plain_scalar_value T, typename N>
+EVE_FORCEINLINE wide<T, typename N::split_type>
+                riscv_lmul_trunc(wide<T, N> a) noexcept
```
Can't simd_cast do this now?

ita-sc replied:
In final result we will call the same consequence of intrinsics, as `simd_cast` will call slice, that call trunc. For now I expect this should be used only there, and not in common code.

For common code, I suppose we can be better: there I have restriction that result `lmul` at maximum should be twice smaller that original `lmul`, but for `simd_cast` there is no such restriction. And we can just cast with one intrinsic to needed lmul (but this will be too big function to review with this patch).

DenisYaroshevskiy replied:
I have a problem with duplicating the same intrinsic selection logic twice. Either do it in simd_cast or in slice, don't do it twice.

ita-sc replied:
> same intrinsic selection logic twice

For now it is used only here. Of course, when I will implement a more general algorithm, I will move this functionality there. I can move this to `simd_cast` now, but my concern is that it will result in huge block of untested code (as tests for `simd_cast` not in `unit.arch`).

Added TODO with this comment for now.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/slice.hpp:
```diff
@@ -0,0 +1,187 @@
+
+  std::array<logical<wide<T, typename N::split_type>>, 2> that {lower, upper};
+  return that;
+}
```
Merge these 2 by saying `simd_value`. 

```
EVE_FORCEINLINE auto
slice_(EVE_REQUIRES(rvv_), callable_options auto O const&, simd_value auto x) {
  return std::array{ x.slice(lower_), x.slice(upper_) };
}
```

ita-sc replied:
Done

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/slice.hpp:
```diff
@@ -0,0 +1,187 @@
+  auto bits_slice = a.bits().slice(upper_slice_t {});
+
+  auto neq = bits_slice != 0;
+  return simd_cast(neq, as<logical<wide<T, typename N::split_type>>> {});
```
seems like you don't need the `!= 0`

If you simd_cast to m1 uints, slice and simd_cast back. Instead of doing `bits()`

ita-sc replied:
> If you simd_cast to m1 uints

This method will not work, if you have `N<=8` (So you have some undefined bits even for `char`).

>  simd_cast 

For now I forbid for `simd_cast` casting between `logical` and `wide`. I think this could be added when I will create an MR with optimization for logical insert/extract (for now, as discussed, we will use memory to do this)

DenisYaroshevskiy replied:
So - if `n < 8` - do a bit shift, if `n >= 8` -> cast and slice?
I'm ok if you don't want to do the logical simd cast now - put a todo here please.

ita-sc replied:
Added TODO for now

ita-sc commented:
@jfalcou, @DenisYaroshevskiy thanks for adding CI for RISC-V. I've updated MR and add `unit.arch` tests.

DenisYaroshevskiy commented on include/eve/arch/riscv/top_bits.hpp:
```diff
@@ -0,0 +1,145 @@
+    using u_type = detail::make_integer_t<byte_size, unsigned>;
+    auto dirty_value = std::bit_cast<u_type>(storage);
+    if constexpr (static_bits_size >= 8)
+      return dirty_value;
```
I don't think dirty value here is a correct description.
top bits 0 out garbage bits. 
You do that as well: `operator&=(top_bits(ignore_none_ {}));`
So - this is actually a perfectly fine value.

ita-sc replied:
You mean I need to rename? Changed to `raw_value`, seems this fits better

DenisYaroshevskiy commented on include/eve/arch/riscv/top_bits.hpp:
```diff
@@ -0,0 +1,145 @@
+  EVE_FORCEINLINE constexpr auto as_int() const
+  requires(static_bits_size <= 64)
+  {
+    constexpr unsigned byte_size = (static_bits_size + 7) / 8;
```
I'd suggest: `std::max(1, static_bits_size / 8);` I think it's a bit clearer as to what's happening.

ita-sc commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,303 @@
+                  "[eve riscv] - Type is not usable in SIMD register (too big)");
+    constexpr bool int_like   = std::is_integral_v<Type> && std::is_signed_v<Type>;
+    constexpr bool uint_like  = std::is_integral_v<Type> && std::is_unsigned_v<Type>;
+    constexpr bool float_like = std::is_floating_point_v<Type>;
```
Done

ita-sc commented on include/eve/arch/riscv/as_register.hpp:
```diff
@@ -0,0 +1,303 @@
+  }
+  using type = decltype(find());
+  static_assert(!std::is_void_v<type>,
+                "[eve riscv] - Type is not usable in a SIMD register (unknown type)");
```
Done

ita-sc commented on include/eve/arch/riscv/tags.hpp:
```diff
@@ -0,0 +1,75 @@
+//================================================================================================
+template<std::size_t VectorRegisterBitSize> struct rvv_abi_
+{
+  static constexpr std::size_t MaxLmul         = 8;
```
Fixed

DenisYaroshevskiy commented on include/eve/arch/riscv/rvv_common_masks.hpp:
```diff
@@ -0,0 +1,40 @@
+
+template<arithmetic_scalar_value T, typename N>
+EVE_FORCEINLINE logical<wide<T, N>>
+                rvv_none()
```
Probably `false` if the other is true?

ita-sc replied:
Done

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+extract_(EVE_REQUIRES(rvv_), O const&, logical<wide<T, N>> v, std::size_t i) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  constexpr size_t                      array_size = (N::value + 7) / 8;
```
N is power of 2. always. you don't need these tricks.

ita-sc replied:
This is needed not to be save with non-power of two. I keep this to support cases when n equals 1, 2, 4. In that case I expect the size will be 1. Changed to std::max.

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+}
+
+// For riscv logical we can not rely on common algorithm.
+template<callable_options O, typename T, typename N>
```
This function should not be needed, default should do it. Why do you have it?

ita-sc replied:
I've touched this in https://github.com/jfalcou/eve/pull/1920#discussion_r1723584925 .

There is a problem in logicals. 

In common code we this:
https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/subscript.hpp#L36

where `abi_t::is_wide_logical` for RISC-V is `false`, so we fail into memcpy, that is wrong for RISC-V logical (as it is used `value_type`, not bool).

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/subscript.hpp#L47

If I set `abi_t::is_wide_logical` to `true`, I will get errors in common slice funtionality:

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/slice.hpp#L85

we have a wrong slice functions for logicals (as mask_type is wide), and bit_cast from `wide` to `logical` is not expected:

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/detail/function/simd/common/slice.hpp#L89

ita-sc replied:
Never mind, It seems due to refactoring now this should work. Removed, thanks

ita-sc replied:
> Never mind, It seems due to refactoring now this should work. Removed, thanks

Finally fixed. The original problem was as I did not change 

https://github.com/jfalcou/eve/blob/318d0cbf1b0705c900173e6e4a0650d1bed2dbb4/include/eve/arch/cpu/tags.hpp#L55

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+// For riscv logical we can not rely on common algorithm.
+template<callable_options O, typename T, typename N, typename Value>
+EVE_FORCEINLINE void
+insert_(EVE_REQUIRES(rvv_), O const&, logical<wide<T, N>>& p, std::size_t i, Value v) noexcept
```
same here

DenisYaroshevskiy commented on include/eve/detail/function/simd/riscv/subscript.hpp:
```diff
@@ -0,0 +1,112 @@
+        std::convertible_to<bool> auto x) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  constexpr size_t                      array_size = (N::value + 7) / 8;
```
n is power of 2.

ita-sc replied:
Same as https://github.com/jfalcou/eve/pull/1920#discussion_r1746933558

DenisYaroshevskiy commented:
There are more than 200 comments on this pr.
Github takes like like 20 seconds to load it for me.

How do you feel about recreating the pr? Just push as a clean branch.
Yeah - some things will be repeated but we will have less of a history in the way.

ita-sc commented:
> How do you feel about recreating the pr? Just push as a clean branch.

Created: https://github.com/jfalcou/eve/pull/1950

ita-sc commented:
Closing, work continue in https://github.com/jfalcou/eve/pull/1950
